addons/io/org.openhab.io.transport.modbus/src/main/java/org/openhab/io/transport/modbus/internal/ModbusManagerImpl.java

```diff
@@ -636,23 +647,23 @@ private void verifyTaskIsRegistered(PollTask task) throws PollTaskUnregistered {
                                 tryIndex, request, e.getClass().getName(), e.getMessage(), operationId, e);
                     }
                     // Invalidate connection, and empty (so that new connection is acquired before new retry)
-                    invalidate(endpoint, connection);
+                    timer.connection.timeConsumer(c -> invalidate(endpoint, c), connection);
                     connection = Optional.empty();
                     continue;
                 } finally {
                     lastTryMillis = System.currentTimeMillis();
                     // Connection was reseted in error handling and needs to be reconnected.
                     // Try to re-establish connection.
                     if (willRetry && !connection.isPresent()) {
-                        connection = getConnection(operationId, oneOffTask, task);
+                        connection = getConnection(timer, oneOffTask, task);
                     }
                 }
             }
             Exception exception = lastError.get();
             if (exception != null) {
                 // All retries failed with some error
                 if (callback != null) {
-                    callbackThreadPool.execute(() -> {
```

Comment:
We want to run in some thread right away, similar to other callbacks

Should have spotted this already in https://github.com/openhab/openhab2-addons/pull/4158